azure/services/natgateways/natgateways.go

```diff
@@ -0,0 +1,123 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package natgateways
+
+import (
+	"context"
+
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2019-06-01/network"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+
+	"sigs.k8s.io/cluster-api-provider-azure/azure"
+	"sigs.k8s.io/cluster-api-provider-azure/azure/scope"
+	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
+)
+
+// NatGatewayScope defines the scope interface for nat gateway service
+type NatGatewayScope interface {
+	logr.Logger
+	azure.ClusterDescriber
+	azure.NetworkDescriber
+	NatGatewaySpecs() []azure.NatGatewaySpec
+}
+
+// Service provides operations on azure resources
+type Service struct {
+	Scope NatGatewayScope
+	client
+}
+
+// New creates a new service.
+func New(scope *scope.ClusterScope) *Service {
+	return &Service{
+		Scope:  scope,
+		client: newClient(scope),
+	}
+}
+
+// Reconcile gets/creates/updates a nat gateway.
+func (s *Service) Reconcile(ctx context.Context) error {
+	ctx, span := tele.Tracer().Start(ctx, "natgateways.Service.Reconcile")
+	defer span.End()
+
+	if !s.Scope.Vnet().IsManaged(s.Scope.ClusterName()) {
+		s.Scope.V(4).Info("Skipping nat gateways reconcile in custom vnet mode")
+		return nil
+	}
+
+	for _, natGatewaySpec := range s.Scope.NatGatewaySpecs() {
+		_, err := s.Get(ctx, s.Scope.ResourceGroup(), natGatewaySpec.Name)
+
+		switch {
+		case err != nil && !azure.ResourceNotFound(err):
+			return errors.Wrapf(err, "failed to get nat gateway %s in %s", natGatewaySpec.Name, s.Scope.ResourceGroup())
+		case err == nil:
+			// nat gateway already exists, leave it alone
+			continue
+		default:
+			// nat gateway was specified in the subnet but it doesn't exist, let's create it
+			s.Scope.V(2).Info("creating Nat Gateway", "nat gateway", natGatewaySpec.Name)
+			err = s.client.CreateOrUpdate(
+				ctx,
+				s.Scope.ResourceGroup(),
+				natGatewaySpec.Name,
+				network.NatGateway{
+					Location: to.StringPtr(s.Scope.Location()),
+					NatGatewayPropertiesFormat: &network.NatGatewayPropertiesFormat{
+						PublicIPAddresses: &[]network.SubResource{
+							{
+								ID: to.StringPtr(natGatewaySpec.NatGatewayIP.Name),
+							},
+						},
+					},
+				},
+			)
```

Comment:
Since we are not using the returned value from Get, can we just not directly do `createOrupdate` and handle the error?

Reply:
I changed the approach a bit, please take a look.